File: ios/ParkPassCore/Sources/ParkPassCore/Visits.swift
import Foundation

/// slug → ISO date (yyyy-mm-dd) of the visit. See docs/specs/storage.md.
public typealias VisitedMap = [String: String]

public struct MergeResult: Equatable, Sendable {
    public let visits: VisitedMap
    public let added: Int
    public let updated: Int

    public var changed: Bool { added > 0 || updated > 0 }
}

public enum Visits {
    /// Union, never a replacement — importing a transfer code must not be
    /// able to destroy pins this device already has. On a collision the
    /// earlier visit wins (ISO dates sort chronologically as plain strings).
    public static func merge(_ current: VisitedMap, _ incoming: VisitedMap) -> MergeResult {
        var next = current
        var added = 0
        var updated = 0
        for (slug, date) in incoming {
            if let existing = next[slug] {
                if date < existing {
                    next[slug] = date
                    updated += 1
                }
            } else {
                next[slug] = date
                added += 1
            }
        }
        return MergeResult(visits: next, added: added, updated: updated)
    }

    /// Same JSON object the web app keeps in localStorage.
    public static func toJSON(_ visits: VisitedMap) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: visits, options: [.sortedKeys]) else { return "{}" }
        return String(decoding: data, as: UTF8.self)
    }

    /// Corrupt or non-object values read as an empty board — never throws.
    public static func fromJSON(_ text: String?) -> VisitedMap {
        guard let text, let data = text.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return [:] }
        return object.compactMapValues { $0 as? String }
    }

    /// Today's *local* calendar date — not UTC, or a pin made just after
    /// midnight Swedish time would be stamped yesterday.
    public static func todayISO(calendar: Calendar = .current, now: Date = Date()) -> String {
        let c = calendar.dateComponents([.year, .month, .day], from: now)
        return String(format: "%04d-%02d-%02d", c.year ?? 1970, c.month ?? 1, c.day ?? 1)
    }
}

File: android/core/src/main/kotlin/com/hopchenko/parkpass/core/Visits.kt
package com.hopchenko.parkpass.core

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull

/** slug → ISO date (yyyy-mm-dd) of the visit. See docs/specs/storage.md. */
typealias VisitedMap = Map<String, String>

data class MergeResult(val visits: VisitedMap, val added: Int, val updated: Int) {
    val changed: Boolean get() = added > 0 || updated > 0
}

object Visits {
    /**
     * Union, never a replacement — importing a transfer code must not be able
     * to destroy pins this device already has. On a collision the earlier
     * visit wins, since that is the one that actually happened first (ISO
     * yyyy-mm-dd sorts chronologically as a plain string).
     */
    fun merge(current: VisitedMap, incoming: VisitedMap): MergeResult {
        val next = current.toMutableMap()
        var added = 0
        var updated = 0
        for ((slug, date) in incoming) {
            val existing = next[slug]
            if (existing == null) {
                next[slug] = date
                added++
            } else if (date < existing) {
                next[slug] = date
                updated++
            }
        }
        return MergeResult(next, added, updated)
    }

    /** Same JSON object the web app keeps in localStorage. */
    fun toJson(visits: VisitedMap): String =
        JsonObject(visits.mapValues { JsonPrimitive(it.value) }).toString()

    /** Corrupt or non-object values read as an empty board — never throws. */
    fun fromJson(text: String?): VisitedMap {
        if (text.isNullOrBlank()) return emptyMap()
        return try {
            val obj = Json.parseToJsonElement(text) as? JsonObject ?: return emptyMap()
            buildMap {
                for ((slug, value) in obj) {
                    val date = (value as? JsonPrimitive)?.takeIf { it.isString }?.contentOrNull
                    if (date != null) put(slug, date)
                }
            }
        } catch (_: Exception) {
            emptyMap()
        }
    }
}
